R/slice.R

```diff
@@ -165,6 +176,96 @@ vec_as_index <- function(i, n, names = NULL) {
   vec_assert(n, integer(), 1L)
   .Call(vctrs_as_index, i, n, names)
 }
+#' @rdname vec_as_index
+#' @export
+vec_as_position <- function(i, n, names = NULL) {
+  if (is.object(i)) {
+    if (vec_is_subtype(i, lgl())) {
+      stop_position_bad_type(i)
+    } else if (vec_is_subtype(i, int())) {
+      i <- vec_cast(i, int())
+    } else if (vec_is_subtype(i, chr())) {
+      i <- vec_cast(i, chr())
+    } else {
+      stop_position_bad_type(i)
+    }
+  } else if (is_double(i)) {
+    i <- vec_coercible_cast(i, int())
+  }
+
+  type <- typeof(i)
+  if (!type %in% c("integer", "character") ||
+      length(i) != 1L ||
+      is.na(i) ||
+      (type == "integer" && i < 1L)) {
+    stop_position_bad_type(i)
+  }
+
+  vec_as_index(i, n, names = names)
+}
+
+stop_position_bad_type <- function(i) {
+  # Should we derive from `stop_incompatible_type()` once we have
+  # union types? The index is incompatible with `union<chr(), int()>`.
+  abort("", "vctrs_error_position_bad_type", i = i)
+}
+
+#' @export
+conditionMessage.vctrs_error_position_bad_type <- function(c) {
+  i <- c$i
+  arg <- c$arg %||% "i"
+
+  lead <- "Must extract with a single position or name."
+
+  if (is.object(i) || !typeof(i) %in% c("integer", "character")) {
+    type <- vec_ptype_full(c$i)
+    return(glue_lines(
+      lead,
+      glue_error_bullets(
+        i = "Positions and names must be integer or character.",
+        x = "`{arg}` has the wrong type `{type}`."
+      )
+    ))
+  }
+
+  if (length(i) != 1L) {
+    size <- length(c$i)
+    return(glue_lines(
+      lead,
+      glue_error_bullets(
+        i = "Positions and names must be size 1.",
+        x = "`{arg}` has the wrong size {size}."
+      )
+    ))
+  }
+
+  if (is.na(i)) {
+    return(glue_lines(
+      lead,
+      glue_error_bullets(
+        i = "Positions and names can't be missing.",
+        x = "`{arg}` can't be `NA`."
+      )
+    ))
+  }
+
+  if (i < 1L) {
+    return(glue_lines(
+      lead,
+      glue_error_bullets(
+        i = "Positions must be positive integers.",
+        x =
+          if (i == 0L) {
+            "`{arg}` can't be zero."
+          } else {
+            "`{arg}` has the wrong sign: {i}."
```

Comment:
Are you trying to use `:` whenever you say `has the wrong ___:`? If so, it is missing for `has the wrong type` and `has the wrong size`.

Reply:
This branch is different because it shows the whole value, which is why I used `:`

Reply:
Other ideas:

```
✖ `i` (with value -1) has the wrong sign.
✖ `i` has the wrong sign. Its value is -1.
```